src/main/java/hudson/plugins/git/GitSCM.java

```diff
@@ -1068,16 +1079,19 @@ public void checkout(Run<?, ?> build, Launcher launcher, FilePath workspace, Tas
 
         try {
           checkoutCommand.execute();
-        } catch(GitLockFailedException e) {
+        } catch (GitLockFailedException e) {
             // Rethrow IOException so the retry will be able to catch it
             throw new IOException("Could not checkout " + revToBuild.revision.getSha1String(), e);
         }
 
-        build.addAction(new GitTagAction(build, workspace, revToBuild.revision));
+        // Don't add the tag and changelog if we've already processed this BuildData before.
+        if (!buildDataAlreadyPresent) {
+            build.addAction(new GitTagAction(build, workspace, revToBuild.revision));
 
-        if (changelogFile != null) {
-            computeChangeLog(git, revToBuild.revision, listener, previousBuildData, new FilePath(changelogFile),
-                    new BuildChooserContextImpl(build.getParent(), build, environment));
+            if (changelogFile != null) {
+                computeChangeLog(git, revToBuild.revision, listener, previousBuildData, new FilePath(changelogFile),
+                        new BuildChooserContextImpl(build.getParent(), build, environment));
+            }
```

Comment:
For the benefit of reviewers (and future cherry picks, etc.) please revert any gratuitous parts of the diff, especially whitespace-only changes.